cmd/compute-domain-controller/node.go

```diff
@@ -0,0 +1,149 @@
+	labelSelector := &metav1.LabelSelector{
+		MatchExpressions: []metav1.LabelSelectorRequirement{
+			{
+				Key:      computeDomainLabelKey,
+				Operator: metav1.LabelSelectorOpExists,
+			},
+		},
+	}
+
+	factory := informers.NewSharedInformerFactoryWithOptions(
+		config.clientsets.Core,
+		informerResyncPeriod,
+		informers.WithTweakListOptions(func(opts *metav1.ListOptions) {
+			opts.LabelSelector = metav1.FormatLabelSelector(labelSelector)
+		}),
+	)
+	informer := factory.Core().V1().Nodes().Informer()
+
+	m := &NodeManager{
+		config:   config,
+		factory:  factory,
+		informer: informer,
+	}
+
+	m.labelCleanupManager = NewCleanupManager[*corev1.Node](informer, getComputeDomain, m.cleanupLabels)
+	return m
+}
+
+func (m *NodeManager) Start(ctx context.Context) (rerr error) {
+	ctx, cancel := context.WithCancel(ctx)
+	m.cancelContext = cancel
+
+	defer func() {
+		if rerr != nil {
+			if err := m.Stop(); err != nil {
+				klog.Errorf("error stopping Node manager: %v", err)
+			}
+		}
+	}()
+
+	m.waitGroup.Add(1)
+	go func() {
+		defer m.waitGroup.Done()
+		m.factory.Start(ctx.Done())
+	}()
+
+	if !cache.WaitForCacheSync(ctx.Done(), m.informer.HasSynced) {
+		return fmt.Errorf("NodeManager: informer cache sync failed")
+	}
+
+	if err := m.labelCleanupManager.Start(ctx); err != nil {
+		return fmt.Errorf("NodeManager: error starting labelCleanupManager: %w", err)
+	}
+
+	return nil
+}
+
+func (m *NodeManager) Stop() error {
+	if err := m.labelCleanupManager.Stop(); err != nil {
+		return fmt.Errorf("NodeManager: error stopping labelCleanupManager: %w", err)
+	}
+	m.cancelContext()
+	m.waitGroup.Wait()
+	return nil
+}
+
+// RemoveComputeDomainLabels() searches all nodes that are currently labeled for
+// a specific CD (as identified via CD UID). It then removes that label from all
+// such nodes.
+func (m *NodeManager) RemoveComputeDomainLabels(ctx context.Context, cdUID string) error {
+	labelSelector := &metav1.LabelSelector{
+		MatchExpressions: []metav1.LabelSelectorRequirement{
+			{
+				Key:      computeDomainLabelKey,
+				Operator: metav1.LabelSelectorOpIn,
+				Values:   []string{cdUID},
+			},
+		},
+	}
+
+	nodes, err := m.config.clientsets.Core.CoreV1().Nodes().List(ctx, metav1.ListOptions{
+		LabelSelector: metav1.FormatLabelSelector(labelSelector),
+	})
+	if err != nil {
+		return fmt.Errorf("error retrieving nodes: %w", err)
+	}
+
+	for _, node := range nodes.Items {
+		newNode := node.DeepCopy()
+		delete(newNode.Labels, computeDomainLabelKey)
+
+		if _, err := m.config.clientsets.Core.CoreV1().Nodes().Update(ctx, newNode, metav1.UpdateOptions{}); err != nil {
+			return fmt.Errorf("error updating node %s: %w", node.Name, err)
+		}
+	}
+
+	if len(nodes.Items) > 0 {
+		var names []string
+		for _, n := range nodes.Items {
+			names = append(names, n.Name)
+		}
+		klog.V(6).Infof("Removed label(s) for CD %v from node(s): %v", cdUID, strings.Join(names, ", "))
+	}
```

Comment:
I think this can be optimized to only delete the label / call update on the node for nodes that actually have the desired label.

Something like (untested):
```
	var names []string
	for _, node := range nodes.Items {
		if node.Labels[computeDomainLabelKey] != cdUID {
			continue
		}
		
		newNode := node.DeepCopy()
		delete(newNode.Labels, computeDomainLabelKey)
		if _, err := m.config.clientsets.Core.CoreV1().Nodes().Update(ctx, newNode, metav1.UpdateOptions{}); err != nil {
			return fmt.Errorf("error updating node %s: %w", newNode.Name, err)
		}
		
		names = append(names, newNode.Name)
	}

	if len(names) > 0 {
		klog.V(6).Infof("Removed label(s) for CD %v from node(s): %v", cdUID, strings.Join(names, ", "))
	}
```

Reply:
I have second thoughts on this: https://github.com/NVIDIA/k8s-dra-driver-gpu/pull/404/files#r2152166284

(yes, let's construct `names` on the fly in the first loop's body; but I hope the `continue` fast path is a noop, in which case we should remove it)

Reply:
yes, safe to remove. see comment in https://github.com/NVIDIA/k8s-dra-driver-gpu/pull/404#discussion_r2152212796